TaskLayer/MbrAnalysis/MbrAnalysisRunner.cs

```diff
@@ -0,0 +1,273 @@
+                    {
+                        ChromatographicPeak mbrPeak = fileSpecificMbrPeaks[i];
+                        PeptideSpectralMatch bestDonorPsm = allPeptides.Where(p =>
+                            p.FullSequence == mbrPeak.Identifications.First().ModifiedSequence).First();
+                        if (bestDonorPsm == null) break;
+                        PeptideWithSetModifications bestDonorPwsm = bestDonorPsm.BestMatchingPeptides.First().Peptide;
+
+                        IEnumerable<PeptideSpectralMatch> peptideSpectralMatches =
+                            mcse.SearchAroundPeak(bestDonorPwsm, mbrPeak.Apex.IndexedPeak.RetentionTime);
+
+                        if (peptideSpectralMatches == null || !peptideSpectralMatches.Any())
+                        {
+                            bestMbrMatches.Add(new MbrSpectralMatch(null, mbrPeak));
+                        }
+                        else
+                        {
+                            bestMbrMatches.Add(new MbrSpectralMatch(BestPsmForMbrPeak(peptideSpectralMatches),
+                                mbrPeak));
+                        }
+                    }
+                });
+            }
+
+            List<PeptideSpectralMatch> allPsms = parameters.AllPsms.
+                OrderByDescending(p => p.Score).
+                ThenBy(p => p.FdrInfo.QValue).
+                ThenBy(p => p.FullFilePath).
+                ThenBy(x => x.ScanNumber).
+                ThenBy(p => p.FullSequence).
+                ThenBy(p => p.ProteinAccession).ToList();
+
+            Directory.CreateDirectory(Path.Join(parameters.OutputFolder, mbrAnalysisFolder));
+            AssignEstimatedPsmQvalue(bestMbrMatches, allPsms);
+            FDRAnalysisOfMbrPsms(bestMbrMatches, allPsms, parameters, fileSpecificParameters);
+            AssignEstimatedPsmPepQValue(bestMbrMatches, allPsms);
+            foreach (MbrSpectralMatch match in bestMbrMatches) match.FindOriginalPsm(allPsms);
+            WriteMbrPsmResults(bestMbrMatches, parameters);
+        }
+
+        private static List<PeptideSpectralMatch> GetAllPeptides(
+            PostSearchAnalysisParameters parameters,
+            CommonParameters commonParameters,
+            List<(string, CommonParameters)> fileSpecificParameters)
+        {
+            List<PeptideSpectralMatch> peptides = new();
+            peptides = parameters.AllPsms.Where(b => b.FullSequence != null).GroupBy(b => b.FullSequence).Select(b => b.FirstOrDefault()).ToList();
+
+            new FdrAnalysisEngine(peptides, parameters.NumNotches, commonParameters, fileSpecificParameters, new List<string> { parameters.SearchTaskId }, "Peptide").Run();
+
+            if (!parameters.SearchParameters.WriteDecoys)
+            {
+                peptides.RemoveAll(b => b.IsDecoy);
+            }
+            if (!parameters.SearchParameters.WriteContaminants)
+            {
+                peptides.RemoveAll(b => b.IsContaminant);
```

Comment:
These might benefit from [PLINQ](https://docs.microsoft.com/en-us/dotnet/standard/parallel-programming/introduction-to-plinq). There's no downside, since "PLINQ analyzes the overall structure of the query" and only performs concurrent tasks if it sees a benefit of parallelization.